### PR TITLE
br: defer not in for-loop in the sendSplitRegionRequest

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #31716

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/restore/split_client.go [hunk 1]
```diff
@@ -294,85 +294,99 @@ func splitRegionWithFailpoint(
 	})
 }
 
+func sendSplitRegionRequest(c *pdClient, ctx context.Context, regionInfo *RegionInfo, keys [][]byte, splitErrors *error, retry int) (bool, *kvrpcpb.SplitRegionResponse, error) {
```
Comment (gozssky) [inline br/pkg/restore/split_client.go:297]: Nit: Moving `func sendSplitRegionRequest` after `func (c *pdClient) sendSplitRegionRequest` may make the diff more easy to review.

File: br/pkg/restore/split_client.go [hunk 2]
```diff
@@ -294,85 +294,99 @@ func splitRegionWithFailpoint(
...
+			resp.RegionError.StaleCommand != nil {
+			log.Warn("a error occurs on split region",
+				zap.Int("retry times", retry),
+				zap.Uint64("regionID", regionInfo.Region.Id),
+				zap.String("error", resp.RegionError.Message),
+				zap.Any("error verbose", resp.RegionError),
+			)
+			return true, nil, nil
+		}
+		return false, nil, nil
```
Comment (gozssky) [inline br/pkg/restore/split_client.go:369]: This return may cause `func (c *pdClient) sendSplitRegionRequest`  returns nil response with nil error which will lead to panic.

Comment (gozssky) [reply br/pkg/restore/split_client.go:369]: I think we can just return `*kvrpcpb.SplitRegionResponse, error`:
nil, nil  => caller need to retry
nil, err => unretryable error
resp, nil => success

Comment (hawkingrei) [reply br/pkg/restore/split_client.go:369]: it is hard. bacause [this](https://github.com/pingcap/tidb/blob/master/br/pkg/restore/split_client.go#L373) need to return all appending error. so It has four condition to implement.

**Review by gozssky — APPROVED** [review]

File: br/pkg/restore/split_client.go [hunk 3]
```diff
@@ -299,82 +299,96 @@ func (c *pdClient) sendSplitRegionRequest(
...
-			return nil, multierr.Append(splitErrors, err)
+		if result != nil {
+			return result, nil
 		}
-		defer conn.Close()
-		client := tikvpb.NewTikvClient(conn)
-		resp, err := splitRegionWithFailpoint(ctx, regionInfo, peer, client, keys, c.isRawKv)
-		if err != nil {
-			return nil, multierr.Append(splitErrors, err)
+		return nil, errors.Trace(splitErrors)
```
Comment (3pointer) [inline br/pkg/restore/split_client.go:312]: seems this line makes for loop not work

Comment (hawkingrei) [reply br/pkg/restore/split_client.go:312]: this logic is the origin site in https://github.com/pingcap/tidb/blob/master/br/pkg/restore/split_client.go#L373

Comment (3pointer) [reply br/pkg/restore/split_client.go:312]: okay

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**